### ISSUE TITLE
HTTP L7 load balancer / reverse proxy

### ISSUE THREAD

**Issue by lavalamp** [kubernetes/kubernetes#561]
Configurable, using labels to route traffic.

From discussion: https://groups.google.com/d/msg/google-containers/frOLMyNl5U4/W5_DQUL933IJ

>>I suppose, if we had some sort of dynamic router based on label queries, you could make that work for your needs with a bit of configuration. I'm not sure if there's really a need for that once DNS naming is set up, though.
 
>I think you're going to want to incorporate some sort of http/s router or at least have a suggested means of configuring one. It seems to be one of the most obvious use cases for Kubernetes.

Like I said, I'm not sure this is needed if there's a good load balancer and DNS name resolution, but filing this for tracking the discussion.

**Comment by phemmer** [comment]
This is very much of interest to us as we run numerous applications, all fronted by a single external endpoint (eg, http://api.example.com).

We've run our own in-house layer 7 load balancer / router for a few years now, and aside from URL rewriting (which is mentioned with the 'target path' thing), the only other feature I can think of which would be of interest is source address filtering.
Our specific use case for this is to allow internal applications to talk to other internal applications through the load balancer using 'private' routes. But we could probably work around this by just adding authentication to these routes.

I'm also wondering if it would be good to define what http headers get added to the request. Since a layer 7 router hides a lot of the client information, this'll need to be passed on. Information such as client IP address, client/server port, whether the client is using SSL, SSL client cert subject & verification status, etc.

**Comment by supirman** [comment]
Hi, I am master student who want to apply to GSOC. I am interested with this problem. And I had project  that involving nginx as load balancer and reverse proxy in the past. Can this be implemented using nginx?

**Comment by smarterclayton** [comment]
It could, although we would prefer a solution that is generic to load balancers.  The OpenShift route concept and mechanism is mostly fully implemented and integrates now with HAProxy and F5 automatically, and has a generic template mode for any other router.  I think that a lot of the work for this has really been done, and it's a matter of defining how we evolve the service api and then moving that code over to Kube.

**Comment by glerchundi** [comment]
Hi guys, I've already created a working example of http reverse proxy and loadbalancer inside kubernetes using nginx + confd (with etcd backend, which is just a proxy to the master etcd). It is composed by three components:
* etcd-proxy: it persists routing and upstreams data
* [nginx-loadbalancer](https://github.com/just-containers/nginx-loadbalancer): using nginx and confd, it works as a reverse proxy load balancing traffic based in the registered upstreams.
* [loadbalancer-feeder](https://github.com/just-containers/loadbalancer-feeder): Listens to kubernetes pod events using [kubelistener](https://github.com/glerchundi/kubelistener) (single ones or created by replication controllers) and updates loadbalancer upstreams accordingly.

A working controller+service example is also available at: https://github.com/glerchundi/kubernetes-http-loadbalancer

Any comment is really appreciated!

**Comment by bgrant0607** [comment]
@glerchundi Thanks for the pointers! Just a quick comment for now since we're trying to wrap up [1.0](https://github.com/GoogleCloudPlatform/kubernetes/issues?q=is%3Aopen+is%3Aissue+milestone%3Av1.0). If you define a service (set portalIP: "None" if you don't need a VIP allocated) for your pods, the endpoints controller will generate a list of their addresses and ports in an Endpoints object for you.
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/services.md#headless-services

**Comment by glerchundi** [comment]
@bgrant0607 ok, thanks for pointing that out. After watching some videos about kubernetes, it seems that the best way/pattern to handle this is to watch services and not replication controllers / pods and modify the http load-balancer accordingly.

Something like:

Service 1: App Production (label: app, production)
Service 2: App Canary (label: app, canary)
Service 3: Load Balancer (selector: app)

This would balance between two services (App Production & App Canary) which in turn will balance between all pods belonging to the corresponding replication controller.

**Comment by bgrant0607** [comment]
@glerchundi Do you mean "service 1" and "service 2" or "replication controller 1" and "replication controller 2"? Otherwise, yes, that's the recommended approach.

**Comment by glerchundi** [comment]
I mean:
```
  ServiceA (App - Prod for lisa.contoso.com) -> RepCtrlA -> PodA{1,2,3,...}
  ServiceB (App - Canary for lisa.contoso.com) -> RepCtrlB -> PodB{1,2,3,...}
  ServiceC (Nginx LB) -> RepCtrlC -> PodC{1,2,3,...}
```

So that nginx, could watch to more than once service (based on a selector) and converge them into just one host. For example in this case the way down will be as follows:
* customer hits nginx from the outside asking for lisa.contoso.com
* kubernetes will balance request (ServiceC) to one of the pods belonging to RepCtrlC.
* nginx will check its reverse proxy configuration for `Host: lisa.contoso.com` (which it was automatically generated using [`kubelistener`](https://github.com/glerchundi/kubelistener) and [`loadbalancer-feeder`](https://github.com/just-containers/loadbalancer-feeder)) and forwards the request to one of those services (ServiceA or ServiceB) which in turn, the chosen service, will balance to the destination pod using its concrete replication controller.
* and finally it will be processed and returned to the end-user.

@bgrant0607 WDYT?

**Comment by erictune** [comment]
That seems like a fine approach if your goal is to send a percentage of
traffic to services that that are slightly different (A/B testing, etc).
In that case, the set of backend services is limited to 2 or a small number.

I assumed that the initial issue was for doing L7 load balancing across
identical replicas of a possibly large and varying number.  In which case,
the pattern you mention doesn't work well.

**Comment by bprashanth** [comment]
In openshift you have a kubernetes service, a set of routes, and an admin created router. The router hooks up the service to its routes. 

The way we currently handle L3 spins this model around because the service declares that it wants a loadbalancer up front, and each service gets a different lb. 

I initially entertained the idea of extending our model for its simplicity, but it breaks down for more complex scenarios. Here's a quick stab at some things I feel are important in the long run:

* We should be able to associate different host+url endpoints (routes) of the same host with different loadbalancers and security configurations. I want to be able to say, /qaService uses http on a haproxy lb and /externalService uses a gce l7 with these certs. Embedding them in the service will get messy.

* Each route should be able to request (initiate the provisioning of) or join an existing loadbalancer. It shouldn't really have to care either way. This is like the persistent claims model where an admin declares different classes of loadbalancers and the route just asks to join one. 

So framing this in the context of 1.1, I propose we implement support for the most basic type of route:

```yaml
type: route
service: svc
host: foo.bar.com
paths:
  - /prod
  - /test
secret: cert
```

1. Every route gets a new gce loadbalancer (much like our l4 model). 
2. The 1:1 mapping simplifies the security configuration. If a secret is mentioned, it's used for tls termination.
3. This route will create an l4 instead of an l7 if didn't have paths. 
4. Services that want https passthrough get an l4. 
5. This doesn't address fan-in, because a route can only have 1 host<->service mapping.

There is a lot of room for optimization, I'm going for the smallest subset of features that will given us a useful l7 api for 1.1.

In parallel (but not necessarily for 1.1) we can develop the claims-plugin model. With that routes can join an existing lb through a claim, and we get fanin and the ability to use multiple certs on a single lb on backends that support it. 

@pweil- @smarterclayton wdyt?

**Comment by smarterclayton** [comment]
On Aug 5, 2015, at 9:51 PM, Prashanth B <notifications@github.com> wrote:

In openshift you have a kubernetes service, a set of routes, and an admin
created router. The router hooks up the service to its routes.

The way we currently handle L3 spins this model around because the service
declares that it wants a loadbalancer up front, and each service gets a
different lb.

I initially entertained the idea of extending our model for its simplicity,
but it breaks down for more complex scenarios. Here's a quick stab at some
things I feel are important in the long run:

   -

   We should be able to associate different host+url endpoints (routes) of
   the same host with different loadbalancers and security configurations. I
   want to be able to say, /qaService uses http on a haproxy lb and
   /externalService uses a gce l7 with these certs. Embedding them in the
   service will get messy.
   -

   Each route should be able to request (initiate the provisioning of) or
   join an existing loadbalancer. It shouldn't really have to care either way.
   This is like the persistent claims model where an admin declares different
   classes of loadbalancers and the route just asks to join one.


There's also the case where the admin decides what loadbalancer gets the
result via some out of band process (routers claim routes via their own
decisions)



So framing this in the context of 1.1, I propose we implement support for
the most basic type of route:

type: routeservice: svchost: foo.bar.compaths:
  - /prod
  - /testsecret: cert


   1. Every route gets a new gce loadbalancer (much like our l4 model).


I think this should be an easily configured admin choice - that's the best
of both worlds  - choose to bind routes to new lbs if a flag on controller
manager says so, or similar.


   1. The 1:1 mapping simplifies the security configuration. If a secret is
   mentioned, it's used for tls termination.
   2. This route will create an l4 instead of an l7 if didn't have paths.
   3. Services that want https passthrough get an l4.
   4. This doesn't address fan-in, because a route can only have 1
   host<->service mapping.



There is a lot of room for optimization, I'm going for the smallest subset
of features that will given us a useful l7 api for 1.1.

In parallel (but not necessarily for 1.1) we can develop the claims-plugin
model. With that routes can join an existing lb through a claim, and we get
fanin and the ability to use multiple certs on a single lb on backends that
support it.


I'd potentially suggest we not have claims for routes, but instead let
routers claim routes.  A route is a claim (effectively) though so I don't
think it's conflicting.

One change we're making right now is to have the spec be the requested host
and the claim/bind/allocate process to set the status field to the
hostnames the route has - this allows two things

1. Route defaulting (new route named foo has empty host and is assigned to
admin configured domain foo.mywildcard.com)
2. First come semantics for hosts on shared routers - users shouldn't be
able to claim a host already claimed.

The important bit is that clients use status of the route to see the
value(s) for the DNS names and ip addresses, or whether the route failed to
bind, vs building in an assumption you can claim host.

@pweil- <https://github.com/pweil-> @smarterclayton
<https://github.com/smarterclayton> wdyt?

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128207917>
.

**Comment by smarterclayton** [comment]
@liggitt @ramr

On Aug 5, 2015, at 9:51 PM, Prashanth B <notifications@github.com> wrote:

In openshift you have a kubernetes service, a set of routes, and an admin
created router. The router hooks up the service to its routes.

The way we currently handle L3 spins this model around because the service
declares that it wants a loadbalancer up front, and each service gets a
different lb.

I initially entertained the idea of extending our model for its simplicity,
but it breaks down for more complex scenarios. Here's a quick stab at some
things I feel are important in the long run:

   -

   We should be able to associate different host+url endpoints (routes) of
   the same host with different loadbalancers and security configurations. I
   want to be able to say, /qaService uses http on a haproxy lb and
   /externalService uses a gce l7 with these certs. Embedding them in the
   service will get messy.
   -

   Each route should be able to request (initiate the provisioning of) or
   join an existing loadbalancer. It shouldn't really have to care either way.
   This is like the persistent claims model where an admin declares different
   classes of loadbalancers and the route just asks to join one.

So framing this in the context of 1.1, I propose we implement support for
the most basic type of route:

type: routeservice: svchost: foo.bar.compaths:
  - /prod
  - /testsecret: cert


   1. Every route gets a new gce loadbalancer (much like our l4 model).
   2. The 1:1 mapping simplifies the security configuration. If a secret is
   mentioned, it's used for tls termination.
   3. This route will create an l4 instead of an l7 if didn't have paths.
   4. Services that want https passthrough get an l4.
   5. This doesn't address fan-in, because a route can only have 1
   host<->service mapping.

There is a lot of room for optimization, I'm going for the smallest subset
of features that will given us a useful l7 api for 1.1.

In parallel (but not necessarily for 1.1) we can develop the claims-plugin
model. With that routes can join an existing lb through a claim, and we get
fanin and the ability to use multiple certs on a single lb on backends that
support it.

@pweil- <https://github.com/pweil-> @smarterclayton
<https://github.com/smarterclayton> wdyt?

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128207917>
.

**Comment by liggitt** [comment]
I'd want to be a little more explicit about the desired tls mode of the route:

1. passthrough
2. terminate (optional cert/key if needed for the requested host)
3. re-encrypt (optional cert/key if needed for the requested host, optional CA bundle for establishing the re-encrypt)

**Comment by bprashanth** [comment]
> I think this should be an easily configured admin choice - that's the best
of both worlds  - choose to bind routes to new lbs if a flag on controller
manager says so, or similar.

SG.

>I'd potentially suggest we not have claims for routes, but instead let
routers claim routes.  A route is a claim (effectively) though so I don't
think it's conflicting.

Who creates the routers? This sounds like a more admin centric approach. How do we detect route overload (or more realistically, decryption saturating cpu)? I was trying to give users the responsibility of deciding weather or not they want a new router. The admin just configures the tiers of available routers (i.e production -> gce, qa -> nginx etc). 

> One change we're making right now is to have the spec be the requested host
and the claim/bind/allocate process to set the status field to the
hostnames the route has - this allows two things
1. Route defaulting (new route named foo has empty host and is assigned to
admin configured domain foo.mywildcard.com)
2. First come semantics for hosts on shared routers - users shouldn't be
able to claim a host already claimed.

This is actually another layer over l7 IMO. Why put route defaulting in the router instead of the apiserver? I guess I don't understand the requirement here, is it actually per router defaulting? or is the per router part just a means to an end.

> I'd want to be a little more explicit about the desired tls mode of the route:
passthrough
terminate (optional cert/key if needed for the requested host)
re-encrypt (optional cert/key if needed for the requested host, optional CA bundle for establishing the re-encrypt)

Are you asking for a new tlsMode field vs having the router infer the mode based on the secret? If so, what's the requirement here: If I say re-encrypt but don't specify a secret, just generate one? I was trying to avoid the cognitive friction of understanding these modes, but a field that defaults to Termination if a secret is specifed will also achieve the same as my initial spec, I suppose.

**Comment by liggitt** [comment]
I think mode should be explicit. There are actually four modes:

1. non-tls
2. passthrough
3. terminate
4. re-encrypt

deriving the mode from the presence or absence of a provided secret isn't straightforward

**Comment by pweil-** [comment]
> One change we're making right now is to have the spec be the requested host
> and the claim/bind/allocate process to set the status field to the
> hostnames the route has - this allows two things

I think there is benefit to this model, particularly in the scenario where a route is requesting or being forced to join an existing lb.  If for any reason the lb decides that it cannot support the route (duplication, it's an l4 only lb when l7 options are requested, etc) the route can easily be described to see why it wasn't deployed.

> I think mode should be explicit.

When a single lb is being used to support many types of routes the templates must be making decisions on how to setup frontend and backend configurations (http vs tcp mode, termination semantics, affinity options).  In a lb-per-route scenario a correct template can be chosen for the lb that ONLY supports that route type.

**Comment by smarterclayton** [comment]
The problem is that it depends on whether the router supports that.  So in
a claims model you might ask for passthrough and get it because it's a
singleton TCP loadbalancer, or get it because the router is HAProxy and can
do it via SNI.  Terminate might not be supported by the load balancer - but
it may not be enough to know that the user wanted termination for its own
sake.

Need to think through some aspects of it for sure.  I agree there should be
some explicit mode.

On Thu, Aug 6, 2015 at 2:56 PM, Jordan Liggitt <notifications@github.com>
wrote:

> I think mode should be explicit. There are actually four modes:
>
>    1. non-tls
>    2. passthrough
>    3. terminate
>    4. re-encrypt
>
> deriving the mode from the presence or absence of a provided secret isn't
> straightforward
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128473740>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by bprashanth** [comment]
Updated the proposed resource https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128207917

What's the motivation behind Openshift's RoutePort, can't a route just target a servcie:targetPort directly and abstract everything else? Is there a reason to include it in the initial route resource?

**Comment by smarterclayton** [comment]
On Aug 6, 2015, at 10:28 PM, Prashanth B <notifications@github.com> wrote:

I think this should be an easily configured admin choice - that's the best
of both worlds - choose to bind routes to new lbs if a flag on controller
manager says so, or similar.

SG.

I'd potentially suggest we not have claims for routes, but instead let
routers claim routes. A route is a claim (effectively) though so I don't
think it's conflicting.

Who creates the routers? This sounds like a more admin centric approach.
How do we detect route overload (or more realistically, decryption
saturating cpu)? I was trying to give users the responsibility of deciding
weather or not they want a new router. The admin just configures the tiers
of available routers (i.e production -> gce, qa -> nginx etc).



I would say both are valuable.  A controller that spawns routers or a
controller loop *in* a router that claims routes.  I think we want to allow
that to be the admin choice.


One change we're making right now is to have the spec be the requested host
and the claim/bind/allocate process to set the status field to the
hostnames the route has - this allows two things
1. Route defaulting (new route named foo has empty host and is assigned to
admin configured domain foo.mywildcard.com)
2. First come semantics for hosts on shared routers - users shouldn't be
able to claim a host already claimed.

This is actually another layer over l7 IMO. Why put route defaulting in the
router instead of the apiserver? I guess I don't understand the requirement
here, is it actually per router defaulting? or is the per router part just
a means to an end.


So the simple options are integrating with external load balancers or
creating one load balancer set per route.  However, there is an infinite
amount of deployment topology complexity beyond that.  We have shared
routers because it's cost effective for admins and can easily integrate
deployment wide policy or auth.  But the next steps are different routers
for different use cases - crossover routers between sections of the
deployment, routers to different networks (Vpn or bridging interfaces or
...), or routers for internal use like caching proxies.

So making routes the request from a user that parts of the infrastructure
manifest gets the proper decoupling to allow all the possible patterns.  We
expect to have many routers, and some routes exposed to multiple (I'm not
opposed to one route per router, but there are cases where you want to
allocate an internal and external DNS entry for testing).  I would draw the
line at multiple hosts or secrets per route for now though.


I'd want to be a little more explicit about the desired tls mode of the
route:
passthrough
terminate (optional cert/key if needed for the requested host)
re-encrypt (optional cert/key if needed for the requested host, optional CA
bundle for establishing the re-encrypt)

Are you asking for a new tlsMode field vs having the router infer the mode
based on the secret? If so, what's the requirement here: If I say
re-encrypt but don't specify a secret, just generate one? I was trying to
avoid the cognitive friction of understanding these modes, but a field that
defaults to Termination if a secret is specifed will also achieve the same
as my initial spec, I suppose.


Reencrypt without a secret is a validation error.  Termination doesn't
necessarily require a secret because a shared router can have a default
cert for its wildcard.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128458251>
.

**Comment by smarterclayton** [comment]
On Aug 6, 2015, at 10:37 PM, Jordan Liggitt <notifications@github.com>
wrote:

I'd want to be a little more explicit about the desired tls mode of the
route:

   1. passthrough
   2. terminate (optional cert/key if needed for the requested host)
   3. re-encrypt (optional cert/key if needed for the requested host,
   optional CA bundle for establishing the re-encrypt)


That's right, the router can have a default client cert or just use the ca.
 forgot about that.



—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128401495>
.

**Comment by smarterclayton** [comment]
RoutePort is currently kind of broken - it was done before multiport
services.  The behavior should be to target one or more ports by
identifying what port to target from the route (because you could have 3
ports on your service, or have a headless service).  @liggitt or @pweil-
could say more.

On Thu, Aug 6, 2015 at 5:10 PM, Prashanth B <notifications@github.com>
wrote:

> Updated the proposed resource #561 (comment)
> <https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128207917>
>
> What's the motivation behind Openshift's RoutePort, can't a route just
> target a servcie:targetPort directly and abstract everything else? Is there
> a reason to include it in the initial route resource?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128508244>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by pweil-** [comment]
RoutePort was originally intended to offer the ability to open a custom frontends that routed to your service.  That said, we do need to account for multi-port services.

Custom frontends defined by routes could open up  a can of worms.  While they're possible to implement I think it has the potential to conflict with scheduling unless you can define a rule that only one host networked container can deployed on each node.

**Comment by bprashanth** [comment]
Another idea we discussed was to move from a service-per-route model, to a service-per-path model. Eg:

```yaml
type: route
Spec:
  host: foo.bar.com
  paths:
    - /prod: svc1, port
    - /test: svc2, port
  tlsMode: Termination
  secret: cert
Status:
  host: foo.bar.com
  ingressIp: 134
```

That seems to fit better with my mental model of a website with multiple endpoints serviced by different groups of pods, all sharing a common security policy. 

The openshift model is to have a route for /prod and another for /test, and since they all join the same router things work out. But in a world where a single route creates a new loadbalancer, the ability to specify multiple services per path makes it easier to get a single ip for the entire site.

Is there a reason to handle each path with a different route object?

**Comment by smarterclayton** [comment]
It seems like a different concern - the ability to request a specific
custom edge endpoint can definitely fail in a lot of environments, so we'd
make it a request.  I don't think it's wrong to be able to specify a
constraint.

On Aug 7, 2015, at 3:12 PM, Paul Weil <notifications@github.com> wrote:

RoutePort was originally intended to offer the ability to open a custom
frontends that routed to your service. That said, we do need to account for
multi-port services.

Custom frontends defined by routes could open up a can of worms. While
they're possible to implement I think it has the potential to conflict with
scheduling unless you can define a rule that only one host networked
container can deployed on each node.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128798267>
.

**Comment by smarterclayton** [comment]
I don't think joining is terrible.  The route is an atomic unit of change,
in some cases I'd prefer to change one route from a blue to green
deployment than two.  I think we could make the argument paths together is
strictly better than paths separate.

On Aug 7, 2015, at 8:54 PM, Prashanth B <notifications@github.com> wrote:

Another idea we discussed was to move from a service-per-route model, to a
service-per-path model. Eg:

type: routeSpec:
  host: foo.bar.com
  paths:
    - /prod: svc1, port
    - /test: svc2, port
  tlsMode: Termination
  secret: certStatus:
  host: foo.bar.com
  ingressIp: 134

That seems to fit better with my mental model of a website with multiple
endpoints serviced by different groups of pods, all sharing a common
security policy.

The openshift model is to have a route for /prod and another for /test, and
since they all join the same router things work out. But in a world where a
single route creates a new loadbalancer, the ability to specify multiple
services per path makes it easier to get a single ip for the entire site.


Another wrinkle is that a router might be one or multiple IPs and DNSes.
If we had elastic ip binding, I might want to add that to a given router
(shared or no) and thus there may be multiple effective IPs/DNS entries for
clients.


Is there a reason to handle each path with a different route object?

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128875560>
.

**Comment by bprashanth** [comment]
Ok, so it sounds like the main reason is that a single-service-per-route keeps it hermetic. I mostly buy that. I also like small resources because they're easier to update, watch, display etc, and I can invalidate a single path<->service map (eg: because the service doesn't have nodeport, which is currently required for gce l7). 

So the model we're assuming is: a hostname + multiple url endpoints, each managed by a different kubernetes service. 

To make a useful 1.1 l7 api around this, I think we need the ability to route all requests for that hostname, through a single ip, to different backends based on routes, without creating a global router up front (because gce only allows one cert per loadbalancer ip, and mixing the models complicates things). 

There are 2 ways to achieve this:
1. Keep the routes simple (more like the first route https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128207917) and implement a basic claims model that allows joining 
2. Make the route  expressive enough to accomodate the basic requirement (more like the second route https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128875560), so we have a usable api even without joining 

@thockin and @bgrant0607 wdyt?

**Comment by thockin** [comment]
I'm having a hard time seeing the whole picture from this thread.  I want
to be the voice of "do the simplest thing we can get away with", here.  I
argued with Prashanth that the multitude of tiny route objects feels
awkward to me.  Admittedly, I am not the webbiest guy, but my mental model
is really: Some set of inputs arrive at a mux which decides based on path
which backend Service:Port to send traffic to.

e.g ingress{"foobar.com"} -> map{"/foo": Service{"foo", 80}, "/bar":
Service{"bar", 8080}}

Changing that to multiple route objects seems confusing and unnecessary.
Can someone explain it?

It might help to get a sketch of the data model and some examples using it.

**Comment by smarterclayton** [comment]
I may not have expressed it clearly, but having multiple paths and services per route doesn't seem so bad (nor multiple hosts) because you can change them atomically (which when doing a blue-green cutover has some advantages).  We did not start with that due to caution, but in practice folks have asked for it.

**Comment by thockin** [comment]
Gotcha.  That sort of approximates GCE's API, too.  We do need to think
about what is possible in AWS and others before we design something
unimplementable.

@justinsb (should have looped him in sooner, sorry).

**Comment by smarterclayton** [comment]
If we do have multiples we have to consider partial rejection on shared
routers for duplicate hosts or paths.

On Aug 8, 2015, at 12:15 AM, Tim Hockin <notifications@github.com> wrote:

Gotcha. That sort of approximates GCE's API, too. We do need to think
about what is possible in AWS and others before we design something
unimplementable.

@justinsb (should have looped him in sooner, sorry).

On Fri, Aug 7, 2015 at 9:12 PM, Clayton Coleman <notifications@github.com>
wrote:

> I may not have expressed it clearly, but having multiple paths and
> services per route doesn't seem so bad (nor multiple hosts) because you
can
> change them atomically (which when doing a blue-green cutover has some
> advantages). We did not start with that due to caution, but in practice
> folks have asked for it.
>
> —
> Reply to this email directly or view it on GitHub
> <
https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128903571
>
> .
>

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/561#issuecomment-128904717>
.

**Comment by bprashanth** [comment]
Please review https://github.com/kubernetes/kubernetes/pull/12827 when you have time

**Comment by justinsb** [comment]
AWS ELB has very limited Layer 7 support.  Although it has some Layer 7 features, these are limited to SSL termination (with a single cert), sticky sessions based on cookies, and writing an access log.  There is no path-based routing for example.  Typically you set up ELB in front of nginx/haproxy.  I think we would likely want to do the same thing for AWS, with a k8s managed nginx/haproxy/vulcand.

In other words, the AWS API for load balancing is so limited that I do not think we should even try to constrain the k8s API to fit within it.  Rather, we should have a k8s option that uses a cloudprovider Layer 4 load balancer in front of a k8s managed software load balancer.  If you have a better load balancer (GCE, OpenStack, hardware) then ideally we would allow you to use that instead.  But AWS will be primarily software implemented.

**Comment by phemmer** [comment]
> AWS ELB has very limited Layer 7 support. ... There is no path-based routing for example

In AWS land, they have a separate service for this, [API Gateway](https://aws.amazon.com/api-gateway/). It's basically another layer that sits on top of the ELB.

**Comment by justinsb** [comment]
Oh, good point @phemmer.  I hadn't seen it marketed this way, but it does look like we could indeed use API Gateway as "just" a Layer 7 load balancer.  I can't help but worry that it isn't really what it is intended for, but I'm very happy for the suggestion - we'll have to evaluate it!

**Comment by smarterclayton** [comment]
Yeah, that's how we use ELB today - as the HA layer for a pair of redundant
proxies.

On Mon, Aug 17, 2015 at 10:14 PM, Justin Santa Barbara <
notifications@github.com> wrote:

> Oh, good point @phemmer <https://github.com/phemmer>. I hadn't seen it
> marketed this way, but it does look like we could indeed use API Gateway as
> "just" a Layer 7 load balancer. I can't help but worry that it isn't really
> what it is intended for, but I'm very happy for the suggestion - we'll have
> to evaluate it!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/issues/561#issuecomment-132036743>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by bprashanth** [comment]
>  Rather, we should have a k8s option that uses a cloudprovider Layer 4 load balancer in front of a k8s managed software load balancer. If you have a better load balancer (GCE, OpenStack, hardware) then ideally we would allow you to use that instead. But AWS will be primarily software implemented.

This is exactly the case for loadbalancer classes (either embedded in the ingress point or via claims). I'm a little wary of offer this out of the box, because there are several multi-tier setups (ELB l7 for ssl termination -> nginx, f5 l4 -> apache ssl proxy -> l7 etc). You should have 2 loadbalancer controllers, one for aws and another for haproxy. 

handwaving a bit in this example: 

aws and haproxy loadbalancer controllers running in cluster 
1. create svc
2. create ingresspoint {/foo: svc1, class:haproxy, layer:7}
3. wait for haproxy loadbalancer controller to allocate an ip
4. wrap the ip in a service with nodeport (there have been a couple of discussions on how to do this: https://github.com/kubernetes/kubernetes/issues/10456, https://github.com/kubernetes/kubernetes/issues/1161)
5. create ingresspoint {/foo: nodeportsvc1, class: elb, layer:4}

> In other words, the AWS API for load balancing is so limited that I do not think we should even try to constrain the k8s API to fit within it.

This is why I'd like to move away from the current interface/cloud-provider model, to a more plugin centric approach. Each loadbalancer is a different beast and kube should just get out of the way. Most of the points @justinsb mentioned for ELB are true for GCELB as well.

**Comment by pires** [comment]
/cc @mikedanese 

https://github.com/kubernetes/contrib/tree/master/service-loadbalancer seems a nice project to fork and take further on to support other load-balancing solutions.

**Comment by mikedanese** [comment]
@bprashanth actually authored that package. I'm just git blamed since I moved it out of the main repo.

**Comment by JeanMertz** [comment]
I haven't seen this mentioned here, but is there any thought on integrating the new HTTPS LBs on GCE?

https://cloud.google.com/compute/docs/load-balancing/http/ssl-certificates?hl=en_US

We currently create separate RCs to host a standard nginx reverse-proxy with SSL termination for each "real" service that we host on Kubernetes. This is managable so far, but leveraging Kubernetes to auto-create a managed LB for us, would be much better.

**Comment by bprashanth** [comment]
@pires yeah the real challenge is providing a consistent interface that allows multiple loadbalancers to co-exist in the same cluster. https://github.com/kubernetes/kubernetes/pull/12827/files talks about our efforts in this direction. 

@JeanMertz yes. See same TLS bits on same proposal.

**Comment by jayunit100** [comment]
related to https://github.com/kubernetes/contrib/tree/master/service-loadbalancer which proposes resolution of this issue as a possible next iteration

**Comment by bgrant0607** [comment]
Can this be closed in favor of more specific follow-up issues?

**Comment by thockin** [comment]
I'm closing this in favor of the more detailed bugs, now that we have something.  Yay!
